### PR TITLE
UnorderedMap of Atomic

### PR DESCRIPTION
**Pull request by fuchsto**
Addressing the implications of `dash::Atomic` value semantics, see #289 and #249

### CODE REVIEW

**Comment by fuchsto** [comment]
@fmoessbauer When testing changes to `dash::Atomic`, please use this PR for testing.

With the changes in this PR, using `dash::Atomic` as mapped type in `dash::UnorderedMap` works and modifying access to mapped values lead to atomic operations as expected.

File: dash/include/dash/map/UnorderedMap.h [hunk 1]
```diff
@@ -613,9 +613,10 @@ class UnorderedMap
     /// The element to insert.
     const value_type & value)
   {
-    auto key    = value.first;
-    auto mapped = value.second;
-    DASH_LOG_DEBUG("UnorderedMap.insert()", "key:", key, "mapped:", mapped);
+    auto key = value.first;
```
Comment (devreal) [inline dash/include/dash/map/UnorderedMap.h]: Use reference here, i.e., `auto&`.

File: dash/include/dash/Atomic.h [hunk 2]
```diff
@@ -48,11 +53,10 @@ class Atomic {
 public:
   typedef T value_type;
 
-  Atomic(const Atomic<T> & other) = delete;
+  Atomic() = default;
```
Comment (devreal) [inline dash/include/dash/Atomic.h]: Can be `constexpr`

File: dash/include/dash/Atomic.h [hunk 3]
```diff
@@ -73,6 +77,20 @@ class Atomic {
    */
   operator T() = delete;
 
+  bool operator==(const self_t & other) const {
```
Comment (devreal) [inline dash/include/dash/Atomic.h]: Can be `constexpr` as well.

**Review by devreal — DISMISSED** [review]

File: dash/include/dash/map/UnorderedMap.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/map/UnorderedMap.h]: auto && even

**Review by fmoessbauer — APPROVED** [review]

**Pull request merged**